### PR TITLE
feat(state): Send treestate from non-finalized state to finalized state

### LINKED ISSUE #4824: Send note commitment and history trees from the non-finalized state to the finalized state
## Motivation

When Zebra commits a finalized block, it re-calculates all the note commitment and history trees. This uses a lot of extra CPU. Zebra also fetches these structures from the database for each block coming from the non-finalized state.

## Related Work

This issue is tracked in #3757.

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

When Zebra commits a finalized block, it re-calculates all the note commitment and history trees. This uses a lot of extra CPU. Zebra also fetches these structures from the database for each block coming from the non-finalized state.

The non-finalized state computes the treestate for each block, and it also stores it in RAM. It could pass the treestate over to the finalized state.

## Solution

I created a new structure `FinalizedWithTrees` that wraps the
treestate and the finalized block. I did that because the original
`FinalizedBlock` is `Eq`, but `HistoryTree` can't be `Eq`.

This PR makes Zebra faster because:
1. The finalized state doesn't retrieve the treestate from the disk if
the non-finalized state supplies it.

2. The finalized state doesn't recompute the treestate if the
non-finalized state supplies it.

Close #4824.

### CODE REVIEW

**Comment by teor2345** [comment]
I'm not quite sure how this happened, but it looks like this ticket got transformed into a PR?
(I'm seeing comments from a ticket from last week, and code for the PR that was just opened.)

File: zebra-state/src/service/finalized_state.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state.rs:339]: Nitpick: We could pass a `Treestate` to this function instead, to reduce the number of arguments?

File: zebra-state/src/service/finalized_state/zebra_db/block.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/block.rs:330]: Nitpick: We could pass a `Treestate` to this function instead, to reduce the number of arguments?

File: zebra-state/src/service/non_finalized_state.rs
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state.rs]: Suggestion: We could add a `FinalizedWithTrees::new()` method that takes the block and trees, and returns a `FinalizedWithTrees`.

This helps reduce the length of `finalize()`, and might reduce repeated code in future.

**Review by teor2345 — APPROVED** [review]
This looks great, and it's a lot simpler than I thought!

I'm going to admin-merge this PR now, so other developers can get accurate timings, and we can move on with ticket #4823.

I made some optional code cleanup suggestions, feel free to submit them in another PR.

**Comment by teor2345** [comment]
Actually I'm going to wait for CI before I merge this!

**Comment by upbqdn** [comment]
> I'm not quite sure how this happened, but it looks like this ticket got transformed into a PR?
(I'm seeing comments from a ticket from last week, and code for the PR that was just opened.)

You're right. I tried a new feature in Emacs: `create pull-request from issue`. I thought nothing happened because it seemed like there was no "new" PR. What actually happened was that the issue really got converted into this PR. I won't do that anymore.

I'll open a new ticket that matches this one, and I'll edit the description of this PR.

**Comment by teor2345** [comment]
> I tried a new feature in Emacs: `create pull-request from issue`. I thought nothing happened because it seemed like there was no "new" PR. What actually happened was that the issue really got converted into this PR. I won't do that anymore.
> 
> I'll open a new ticket that matches this one, and I'll edit the description of this PR.

Thanks, it's not a big deal, I mainly commented because I thought it might confuse other developers.
(I can't imagine why GitHub has that feature, it looked like a bug to me!)

**Comment by upbqdn** [comment]
> I'm going to admin-merge this PR now, so other developers can get accurate timings, and we can move on with ticket https://github.com/ZcashFoundation/zebra/issues/4823.

> Actually I'm going to wait for CI before I merge this!

All regular tests pass locally, but I didn't add any new tests. I'm running a full sync now. I'm currently at block 500086. I'd like to wait until tomorrow and see if I can sync Zebra with this PR. I think it might be good to wait until that with the merge as well to make sure we can sync.

**Comment by upbqdn** [comment]
> > I tried a new feature in Emacs: `create pull-request from issue`. I thought nothing happened because it seemed like there was no "new" PR. What actually happened was that the issue really got converted into this PR. I won't do that anymore.
> > I'll open a new ticket that matches this one, and I'll edit the description of this PR.
> 
> Thanks, it's not a big deal, I mainly commented because I thought it might confuse other developers. (I can't imagine why GitHub has that feature, it looked like a bug to me!)

At least we know it's possible.

**Comment by upbqdn** [comment]
I updated the description of this PR, and created #4824.

I didn't update the doc file https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md because it seems already outdated. Should I refactor it, and make it up to date as much as I can?

**Comment by teor2345** [comment]
> I updated the description of this PR, and created #4824. 

Thanks!

> I didn't update the doc file https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md because it seems already outdated. Should I refactor it, and make it up to date as much as I can?

I'm not sure, I think it depends on what the RFC documenting.

RFCs are meant to be a point-in-time change proposal. We've been updating the state RFC to propose database structure changes, but we haven't really updated anything else. And that seems to work for us.

For most code, the best place to document it is in function and module comments. So I think we can just leave the old code in the RFC as it is.

For designs, we can document them in module comments. If we need to, we can also add designs or diagrams to https://github.com/ZcashFoundation/zebra/tree/main/book/src/dev

File: zebra-state/src/service/finalized_state.rs
Comment (teor2345) [inline zebra-state/src/service/finalized_state.rs]: I think what's happening is that the history tree has already been updated in the non-finalized state, so it is now valid for the next block.

Previously, the finalized state validated the history tree, then updated it.
Now, we're validating it, updating it, and trying to validate it again.

As a quick fix, we can move `check::block_commitment_is_valid_for_chain_history()` to the start of the `None` block above.

(And update the comment to explain why!)

**Review by teor2345 — CHANGES_REQUESTED** [review]
I'm seeing this error testing a rollup of all the current high-priority fixes:
```
The application panicked (crashed).
Message:  expected that errors would not occur when writing to disk or updating note commitment and history trees:
InvalidBlockCommitment(InvalidChainHistoryBlockTxAuthCommitment { expected: [19, 214, 187, 9, 105, 227, 98, 101, 186, 168, 27, 89, 97, 105, 131, 27, 62, 52, 102, 174, 118, 81, 84, 222, 113, 95, 99, 77, 2, 147, 118, 105], actual: [84, 204, 12, 218, 41, 255, 48, 86, 121, 59, 42, 105, 146, 16, 154, 40, 171, 132, 108, 184, 12, 237, 106, 253, 31, 185, 41, 149, 134, 176, 24, 222] })
Location: zebra-state/src/service.rs:301
Metadata:
version: 1.0.0-beta.12+46.g0804bb6
Zcash network: Mainnet
state version: 25
branch: test-speed
git commit: 0804bb6
commit timestamp: 2022-07-26T01:02:09Z
target triple: x86_64-unknown-linux-gnu
build profile: release
```

File: zebra-state/src/service/non_finalized_state.rs
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state.rs]: These are the trees for the tip block. But we're sending the root block to the finalised state, so we also need to send the root trees. I think this means we need to store a history tree / height index.

It would also be good to move this code into `pop_root()`, which looks up the height of the root block, and drops all the indexes we're not using any more.

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think we need to send the root block trees to the finalised state, not the tip block trees.

File: zebra-state/src/service/non_finalized_state.rs
Comment (upbqdn) [reply zebra-state/src/service/non_finalized_state.rs]: > These are the trees for the tip block. But we're sending the root block to the finalised state, so we also need to send the root trees.

I think you're right!

File: zebra-state/src/service/finalized_state.rs
Comment (upbqdn) [reply zebra-state/src/service/finalized_state.rs]: Done in [50518fb](https://github.com/ZcashFoundation/zebra/pull/4721/commits/50518fb15c56be25b9d38c189a9f5fd91367ad82).

**Comment by teor2345** [comment]
@upbqdn performance improvements are a low priority right now.

Do you think you can finish this PR this week?
Or did you want to close it, leave a note on the ticket, and move on to something else?

**Comment by upbqdn** [comment]
> Do you think you can finish this PR this week?

Yea, I think I can finish it this week.

**Comment by upbqdn** [comment]
I rebased this PR onto #4928, and I'm using the correct treestate now.
I resolved all issues, so the PR is ready for another review.
I'm running a fresh sync on mainnet to see how it performs.

File: zebra-state/src/service/non_finalized_state.rs
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state.rs:110]: Thanks for this rename!

**Review by teor2345 — APPROVED** [review]
Thanks, looks good!

This also contains most of the work we need to skip rebuilding the treestates on a chain fork, which is our longest state update at the moment. (It's still a low priority, but it will be easy to do later.)

**Comment by upbqdn** [comment]
I could sync Zebra with this PR, but it still took ~15 hours, so no speedup.

**Comment by teor2345** [comment]
> I could sync Zebra with this PR, but it still took ~15 hours, so no speedup.

I don't expect to see a speedup until we finish checkpointing, because this PR only affects full verification.
Is the full verification stage faster?

(We can do some test runs in CI once the critical CI fixes merge, they split out checkpointing and full verification into different GitHub actions jobs.)

**Comment by upbqdn** [comment]
>  Is the full verification stage faster?

From observing the output log, the last two percent of the sync were still pretty slow.

**Comment by teor2345** [comment]
That's interesting, I think I'll run a full sync with and without this PR, and see what the timings are. I just need to check it's not slower.

It's possible there are multiple things slowing down Zebra. So we might just want to merge this PR and come back to performance later, when it's a priority again.

**Review by teor2345 — CHANGES_REQUESTED** [review]
This PR needs a full sync test before it can merge, and we probably want to wait for other performance fixes to merge first.

Edit: we need to wait for PR #4933 to merge, because it changes the performance of full syncs.

Edit 2: we need to wait for #4926 to merge, it will have performance changes. Maybe #5032 will as well.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by upbqdn** [comment]
@mergifyio update

**Comment by teor2345** [comment]
Sorry this is taking a while, but we can do the full syncs here once #4926 merges, because it will have performance changes.

Maybe #5032 will as well, but maybe not, so I think we'd be ok either way.

**Comment by upbqdn** [comment]
On Friday, I started a full-sync using beta 14 with the following times:

- from genesis to block 1705779 (~95.019% of the block chain): 5 hours and 21 minutes
- from block 1705779 to block 1795348: 14 hours and 12 minutes.

I'll post sync times for this PR tomorrow, I'm at 97% at the moment. I had to restart the sync after 15 hours, unfortunately.

**Comment by teor2345** [comment]
> On Friday, I started a full-sync using beta 14 with the following times:
> 
>     * from genesis to block 1705779 (~95.019% of the block chain): 5 hours and 21 minutes
> 
>     * from block 1705779 to block 1795348: 14 hours and 12 minutes.
> 
> 
> I'll post sync times for this PR tomorrow, I'm at 97% at the moment. I had to restart the sync after 15 hours, unfortunately.

This PR only impacts sync times for the non-finalized state, so the most important time will be the "end" job. (The job that comes after the final checkpoint.)

**Comment by upbqdn** [comment]
Here are times for syncing after the last checkpoint, the height of which is 1789180:

This PR:
- from block 1789180 to block 1795401:  1 hour, 20 minutes;
- the height difference is 6221 blocks.

Beta 14:
- from block 1789180 to block 1795405: 2 hours, 20 minutes;
- the height difference is 6225 blocks.

I consider this a noticeable speed-up. 

Edit: the times in the original comment were not correct. I realized I included a few blocks for which beta 14 was waiting in a synced state.

**Review by teor2345 — APPROVED** [review]
Great, thanks for this, and thanks for your patience with this PR.

**Pull request merged**
